Use constexprs where possible

tfry-git opened the pull request:
Work in progress.

Notes to self:
- inverses?
- shifting negative numbers. In particular, is there any case, where we expect -1 >> 1 == 0?

tfry-git commented:
@tomcombriat : This is still a work in progress, but I'd like to get a first look from your side on this, before I continue to apply the same pattern for further operators:

- The first commit (c'tors) is quite straight-forward. I suspect those massive flash-savings are probably just an artifact that is unlikely to turn up in real use (maybe in the Serial.print code).
- The second commit has taken me many hours, today, and this is where I'd really like you to second-check what I did so far. Note that I only adjusted a single operator+ for now, all others would be adjusted according to the same pattern - if it seems correct. See my commit message on https://github.com/tomcombriat/FixMath/pull/4/commits/172610b81ec2f5251f8809d43d64329c7d218380 for some more detail.

Note that once all this is finished, I intend to implement auto-checks using static_asserts(), as outlined, some time ago (while I was still thinking, this was easy to do...).

tomcombriat commented:
Hei!

That looks quite neat! This code is quite hard to read and I think this suggestion actually makes it more readable (I doubt a bit it will ever become ultra readable anyway).

I made some comments but they do not appear in the thread, in case you did not received a notifications, they are [here](https://github.com/tomcombriat/FixMath/commit/172610b81ec2f5251f8809d43d64329c7d218380#comments).

Looking very fast yesterday, I though that you were extensively using `auto` for the signature of the operators (for instance): 
```
constexpr auto operator+ (const UFix<_NI,_NF,_RANGE>& op) const
```
instead of
```
constexpr typename UFix<FixMathPrivate::FM_max(NI,_NI), FixMathPrivate::FM_max(NI,_NI), FixMathPrivate::rangeAdd(NF,_NF,RANGE,_RANGE)>::NIadjusted_t operator+ (const UFix<_NI,_NF,_RANGE>& op) const
```
which I remember to have found very smart as the return type can be inferred by what is returned, but I see now that you removed that. Out of curiosity, why does not this work?

tfry-git commented:
> I made some comments but they do not appear in the thread, in case you did not received a notifications, they are [here](https://github.com/tomcombriat/FixMath/commit/172610b81ec2f5251f8809d43d64329c7d218380#comments).

Replies a the same place. While addressing your points, I also went ahead and added some first static asserts, which would have immediately caught my mixup of NI and NF.
 
> Looking very fast yesterday, I though that you were extensively using `auto` for the signature of the operators (for instance):

[...]

> which I remember to have found very smart as the return type can be inferred by what is returned, but I see now that you removed that. Out of curiosity, why does not this work?

Apparently, this is not available before C++14. Compilation failed on the Uno for that reason.

tfry-git commented:
Ok, so here's the complete PR. We could certainly add a lot more static checks, but I think it's a start.

While doing this, I ran into two issues:
- My test for invAccurate() doesn't work. Am I doing it wrong?
- Turns out, bit-shifts of negative numbers are not formally defined behavior (before C++-20). According to my reading that won't cause a real-world problem, as it has been the de-facto standard on virtually all platforms, but prevents use of anything involving such shifts in constexpr. One small surprise, I came to learn is that ```-2 >> 2 == -1```. I don't think we strongly rely on it being 0 anywhere, though.

tomcombriat commented:
Wow, nice! I'll try to have a closer look at it very soon!

Regarding the issues:

> My test for invAccurate() doesn't work. Am I doing it wrong?

No, this is because of the way the inverse is computed. Let's take a `UFix<4,0> a = 2;`. It's accurate inverse will be a `UFix<0,4>` and is computed by dividing the maximum number that can be hold in the resulting type by the raw value of `a`. Hence, here, it will effectively be computed as (15/4) / 16 (the last division is done by the shifting) and not (16/4)/16 to avoid touching one more bit. It follows that `.invAccurate()` (and all the other inversions) are never exact, even when the resulting number could be represented in the destination type. `.invAccurate()` only guarantees that, in most case, two successive values will not have the same inverse. One way to boost the precision (but never reaching the exact number) is to use `.inv<N>()` which will effectively use more bits for computation. I hope this is more or less clear ;).

> Turns out, bit-shifts of negative numbers are not formally defined behavior (before C++-20).
Yes, but as you said, it is a *de facto* standard that should be present on all compilers this code may concern. Note that, following you comment on the surprise, this is also connected to the `+1` in `typename IntegerType<FixMathPrivate::sBitsToBytes(NI+1)>::signed_type asInt() const` and is caused by two's complement representation if I got everything correct. Basically, the sign bit cannot change with a shift, and this representation does not provide a negative 0, hence shifting a negative number saturates at -1 as you observed. I do not think we should be too concern about this but it is good to keep in mind!

tomcombriat commented on src/FixMath.h:
```diff
@@ -468,55 +455,48 @@ class UFix
-    UFix<new_NI,new_NF> left(*this);
-    UFix<new_NI,new_NF> right(op);
-    return left.asRaw()!=right.asRaw();
+    return !(*this==op);
```
Ah yes, indeed… But I always wonder if `!=` and `==` are equally fast: for the former, once one bit is found different there is no need to go further to know they are different, whereas for the latter the processor needs to check them all. Note that it is likely that it is done in one shot anyway so if an optimization it is, it is probably a micro one…

tfry-git replied:
A valid point. I would hope for this to be optimized, but I am not sure. Since it's not a lot of trouble, I reverted these to the original logic (in new formulation).

tomcombriat commented on src/FixMath.h:
```diff
@@ -742,27 +717,32 @@ class SFix
   }
-  
+
+  /** Addition between a SFix and a UFix. Safe.
```
Did I forgot this one initially? Or maybe I put it as a free function

tfry-git replied:
This and the one or two others used to be free functions. I brought them into the classes to avoid extra friend declarations (for ```U/SFixNIadj_t```).

tomcombriat commented on src/FixMath_Autotests.h:
```diff
@@ -0,0 +1,73 @@
+      // multiplication
+      static_assert(c * UFix<36, 5>(3ll << 31) == UFix<58,0>(33ll*(3ll << 31)));  // NOTE: The exact values are aribrary, but we want something that would overflow the initial type range
+      static_assert(a * UFix<0, 2>(3, true) == UFix<17, 8>(24));  // 32 * .75 == 24
+      // static_assert(a * UFix<5, 0>(4).invAccurate() == UFix<17, 8>(8));  // 32 * (1/4) == 8 // TODO: FAIL. Isn't this supposed to work?
```
See my comment in the thread. This is a limitation because of the way it is computed. Note that we know exactly how far it can be (hence do a static_assert with a range of the output instead of a fixed value).

tfry-git replied:
Ok, I see that now. That cues two further questions (neither of which I'd intend to address within this PR):

- Maybe add a convenience function for ranged comparisons?
- I see what you mean, now, but invAccurate() sounded like - it would be accurate... I guess that's mostly a naming thing. If you don't deem it too late to change that, I'd suggest renaming as perhaps invFull()? And maybe a new function that works closer to what I expected would have a value, too (for small numbers, the difference can become quite noticeable).

tomcombriat replied:
> If you don't deem it too late to change that, I'd suggest renaming as perhaps invFull()? 

Sounds like a good idea! I will try to change that (and add a real accurate) soon. Note that in all cases, if the value cannot be represented in the destination type well, you are done for… If find a bit fun that division are the only operations which lose precision in fixed point math!
I think now is the time for these kind of things, I think I used one or two `invAccurate` in Mozzi (where speed is usually more important than precision) that can easily be changed into `invFull`

tomcombriat commented on src/FixMath_Autotests.h:
```diff
@@ -0,0 +1,73 @@
+
+      // multiplication
+      static_assert(c * UFix<36, 5>(3ll << 31) == UFix<58,0>(33ll*(3ll << 31)));  // NOTE: The exact values are aribrary, but we want something that would overflow the initial type range
+      static_assert(a * UFix<0, 2>(3, true) == UFix<17, 8>(24));  // 32 * .75 == 24
```
I think, as it is considered good practice in the usage of this library, that:
```
static_assert(a * UFix<0, 2>(3, true) == UFix<**5, 0**>(24));
```
is equally correct.

tfry-git replied:
True. That wasn't meant to demonstrate good style, my idea was rather to provoke some shifting across the board.

It might make sense to distill some of these tests into a clean example, esp. to show more clearly how NI and NF get promoted by the various operators.

tomcombriat replied:
I suspected that! Thanks for clarifying!

tomcombriat commented on src/FixMath_Autotests.h:
```diff
@@ -0,0 +1,73 @@
+      static_assert(d.getNF() == 2);     // NF is always max in addition
+      static_assert((a + c).getNF() == 1);
+      static_assert(d.asRaw() == 258);
+      constexpr auto e = d + b + c;      // addition of one 9 (NI) bit and two 8 (NI) bit numbers needs promotion to 10 bits, not 11, only.
```
It was a bit of troubles but I think it is really worth it!

tomcombriat reviewed (approved):
Looks good to me! No testing done but I am working these days on a code that uses `FixMath` extensively so a big mistake won't stay unnoticed for very long.

I added a few comments here and there, mostly to make sure I understand, so feel free to ignore them. 

Once merged, I suggest waiting a couple of weeks to let an obvious bug show its face and then make a new release with that!
